File: android-native/FloatingBubbleService.java
package com.oo.mldraftcounter;

import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.PixelFormat;
import android.os.Build;
import android.os.IBinder;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.view.WindowManager;
import android.widget.LinearLayout;
import android.widget.TextView;

public class FloatingBubbleService extends Service {

    private WindowManager windowManager;
    private View bubbleView;
    private TextView panelText;
    private boolean expanded = false;
    private boolean bubbleAdded = false;

    private SharedPreferences prefs;
    private final SharedPreferences.OnSharedPreferenceChangeListener prefsListener =
        (sp, key) -> {
            if (OverlayBridgePlugin.KEY_TEXT.equals(key)) {
                String text = sp.getString(OverlayBridgePlugin.KEY_TEXT, "Belum ada rekomendasi");
                updateText(text == null ? "" : text);
            }
        };

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        windowManager = (WindowManager) getSystemService(Context.WINDOW_SERVICE);
        prefs = getSharedPreferences(OverlayBridgePlugin.PREFS_NAME, Context.MODE_PRIVATE);
        prefs.registerOnSharedPreferenceChangeListener(prefsListener);

        buildBubbleView();
        addBubbleToWindow();
    }

    private void buildBubbleView() {
        LinearLayout container = new LinearLayout(this);
        container.setOrientation(LinearLayout.VERTICAL);
        container.setBackgroundColor(0xE6121826);
        container.setPadding(24, 16, 24, 16);

        TextView title = new TextView(this);
        title.setText("\u2694 ML Counter");
        title.setTextColor(0xFFE8B64C);
        title.setTextSize(12f);
        title.setPadding(0, 0, 0, 6);

        panelText = new TextView(this);
        panelText.setText("Ketuk untuk buka rekomendasi");
        panelText.setTextColor(0xFFE9EDF5);
        panelText.setTextSize(12f);
        panelText.setMaxWidth(480);

        container.addView(title);
        container.addView(panelText);
        bubbleView = container;
    }

    private void addBubbleToWindow() {
        int overlayType = (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O)
            ? WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY
            : WindowManager.LayoutParams.TYPE_PHONE;

        final WindowManager.LayoutParams params = new WindowManager.LayoutParams(
            WindowManager.LayoutParams.WRAP_CONTENT,
            WindowManager.LayoutParams.WRAP_CONTENT,
            overlayType,
            WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE,
            PixelFormat.TRANSLUCENT
        );
        params.gravity = Gravity.TOP | Gravity.START;
        params.x = 0;
        params.y = 200;

        final int[] initialX = {0};
        final int[] initialY = {0};
        final float[] touchX = {0f};
        final float[] touchY = {0f};
        final boolean[] moved = {false};

        bubbleView.setOnTouchListener((v, event) -> {
            switch (event.getAction()) {
                case MotionEvent.ACTION_DOWN:
                    initialX[0] = params.x;
                    initialY[0] = params.y;
                    touchX[0] = event.getRawX();
                    touchY[0] = event.getRawY();
                    moved[0] = false;
                    return true;
                case MotionEvent.ACTION_MOVE:
                    int dx = (int) (event.getRawX() - touchX[0]);
                    int dy = (int) (event.getRawY() - touchY[0]);
                    if (Math.abs(dx) > 12 || Math.abs(dy) > 12) moved[0] = true;
                    params.x = initialX[0] + dx;
                    params.y = initialY[0] + dy;
                    windowManager.updateViewLayout(bubbleView, params);
                    return true;
                case MotionEvent.ACTION_UP:
                    if (!moved[0]) toggleExpand();
                    return true;
                default:
                    return false;
            }
        });

        windowManager.addView(bubbleView, params);
        bubbleAdded = true;
        String text = prefs.getString(OverlayBridgePlugin.KEY_TEXT, "Belum ada rekomendasi");
        updateText(text == null ? "" : text);
    }

    private void toggleExpand() {
        expanded = !expanded;
        panelText.setVisibility(expanded ? View.VISIBLE : View.GONE);
    }

    private void updateText(String text) {
        panelText.setText((text == null || text.trim().isEmpty()) ? "Belum ada rekomendasi" : text);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        prefs.unregisterOnSharedPreferenceChangeListener(prefsListener);
        if (bubbleAdded && bubbleView != null) {
            windowManager.removeView(bubbleView);
        }
    }
}

File: android-native/FloatingBubbleService.kt
package com.oo.mldraftcounter

import android.app.Service
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.graphics.PixelFormat
import android.os.Build
import android.os.IBinder
import android.view.Gravity
import android.view.MotionEvent
import android.view.View
import android.view.WindowManager
import android.widget.LinearLayout
import android.widget.TextView

/**
 * Bubble mengambang legal (bukan cheat/hack ke game) - mirip chat-head Messenger.
 * Menampilkan ringkasan rekomendasi counter terakhir yang dikirim dari web app
 * lewat OverlayBridgePlugin. Bisa digeser bebas, dan diketuk untuk expand/collapse.
 *
 * WAJIB: pengguna sudah memberi izin "Tampil di atas aplikasi lain"
 * (Settings.canDrawOverlays) sebelum service ini dijalankan.
 */
class FloatingBubbleService : Service() {

    private lateinit var windowManager: WindowManager
    private lateinit var bubbleView: View
    private lateinit var panelText: TextView
    private var expanded = false

    private lateinit var prefs: SharedPreferences
    private val prefsListener = SharedPreferences.OnSharedPreferenceChangeListener { sp, key ->
        if (key == OverlayBridgePlugin.KEY_TEXT) {
            updateText(sp.getString(OverlayBridgePlugin.KEY_TEXT, "Belum ada rekomendasi") ?: "")
        }
    }

    override fun onBind(intent: Intent?): IBinder? = null

    override fun onCreate() {
        super.onCreate()
        windowManager = getSystemService(Context.WINDOW_SERVICE) as WindowManager
        prefs = getSharedPreferences(OverlayBridgePlugin.PREFS_NAME, Context.MODE_PRIVATE)
        prefs.registerOnSharedPreferenceChangeListener(prefsListener)

        buildBubbleView()
        addBubbleToWindow()
    }

    private fun buildBubbleView() {
        val container = LinearLayout(this).apply {
            orientation = LinearLayout.VERTICAL
            setBackgroundColor(0xE6121826.toInt()) // panel gelap semi transparan
            setPadding(24, 16, 24, 16)
        }

        val title = TextView(this).apply {
            text = "⚔ ML Counter"
            setTextColor(0xFFE8B64C.toInt())
            textSize = 12f
            setPadding(0, 0, 0, 6)
        }

        panelText = TextView(this).apply {
            text = "Ketuk untuk buka rekomendasi"
            setTextColor(0xFFE9EDF5.toInt())
            textSize = 12f
            maxWidth = 480
        }

        container.addView(title)
        container.addView(panelText)
        bubbleView = container
    }

    private fun addBubbleToWindow() {
        val overlayType = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O)
            WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY
        else
            WindowManager.LayoutParams.TYPE_PHONE

        val params = WindowManager.LayoutParams(
            WindowManager.LayoutParams.WRAP_CONTENT,
            WindowManager.LayoutParams.WRAP_CONTENT,
            overlayType,
            WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE,
            PixelFormat.TRANSLUCENT
        )
        params.gravity = Gravity.TOP or Gravity.START
        params.x = 0
        params.y = 200

        // Bubble bisa digeser bebas di layar
        var initialX = 0
        var initialY = 0
        var touchX = 0f
        var touchY = 0f
        var moved = false

        bubbleView.setOnTouchListener { _, event ->
            when (event.action) {
                MotionEvent.ACTION_DOWN -> {
                    initialX = params.x
                    initialY = params.y
                    touchX = event.rawX
                    touchY = event.rawY
                    moved = false
                    true
                }
                MotionEvent.ACTION_MOVE -> {
                    val dx = (event.rawX - touchX).toInt()
                    val dy = (event.rawY - touchY).toInt()
                    if (kotlin.math.abs(dx) > 12 || kotlin.math.abs(dy) > 12) moved = true
                    params.x = initialX + dx
                    params.y = initialY + dy
                    windowManager.updateViewLayout(bubbleView, params)
                    true
                }
                MotionEvent.ACTION_UP -> {
                    if (!moved) toggleExpand()
                    true
                }
                else -> false
            }
        }

        windowManager.addView(bubbleView, params)
        updateText(prefs.getString(OverlayBridgePlugin.KEY_TEXT, "Belum ada rekomendasi") ?: "")
    }

    private fun toggleExpand() {
        expanded = !expanded
        panelText.visibility = if (expanded) View.VISIBLE else View.GONE
    }

    private fun updateText(text: String) {
        panelText.text = if (text.isBlank()) "Belum ada rekomendasi" else text
    }

    override fun onDestroy() {
        super.onDestroy()
        prefs.unregisterOnSharedPreferenceChangeListener(prefsListener)
        if (::bubbleView.isInitialized) {
            windowManager.removeView(bubbleView)
        }
    }
    }
